### ISSUE TITLE
Using SAC alone against ART Robust Dpatch attack

### ISSUE THREAD

**Issue by Reza99RD** [joellliu/SegmentAndComplete#1]
hello,
I've applied a patch attack with Robust Dpatch method from ART library [Robust Dpach Example](https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/examples/get_started_fasterrcnn.py) against PyTorch F-RCNN object detector, and then when I'm using SAC alone to detect the patch in the image with the applied patch made by the mentioned method, "x_processed" output image fails to detect and remove the patch. 
can you tell me what am I doing wrong with implementing the code?

I have uploaded image with the adversarial patch.
patch size is (150*150)
Thank you

```
import cv2
import numpy as np
import matplotlib.pyplot as plt
import torch
from patch_detector import PatchDetector

SAC_processor = PatchDetector(3, 1, base_filter=16, square_sizes=[150, 100, 75, 50, 25], n_patch=1)
SAC_processor.unet.load_state_dict(torch.load("ckpts/coco_at.pth", map_location='cpu'))

image_0 = cv2.imread("/content/RDpatch_Adv.jpg")
image_0 = cv2.cvtColor(image_0, cv2.COLOR_BGR2RGB) 

image = np.stack([image_0], axis=0).astype(np.float32)

image=image.transpose(0,3,1,2)
image=torch.tensor(image)

x_processed, _, _ = SAC_processor(image, bpda=True, shape_completion=True)

image_remove = np.asarray(x_processed[0])
image_remove=image_remove.transpose(1,2,0)

plt.axis("off")
plt.figure(figsize = (20,30))
plt.imshow(image_remove.astype(np.uint8), interpolation="nearest")
plt.show()
```

![RDpatch_Adv](https://user-images.githubusercontent.com/117666885/211782958-54fb58af-9c42-47af-82a3-5b6cf56946c8.jpg)

**Comment by joellliu** [comment]
Hi @Reza99RD 

The input image needs to be normalized to [0, 1]. In addition, I would suggest saving the image as png instead of jpg to avoid compression. I tested a similar image and it looks like it is working. Let me know if you have any other questions.

```
import cv2
import numpy as np
import matplotlib.pyplot as plt
import torch
from patch_detector import PatchDetector

SAC_processor = PatchDetector(3, 1, base_filter=16, square_sizes=[150, 100, 75, 50, 25], n_patch=1)
SAC_processor.unet.load_state_dict(torch.load("ckpts/coco_at.pth", map_location='cpu'))

image_0 = cv2.imread("bike.png")
image_0 = cv2.cvtColor(image_0, cv2.COLOR_BGR2RGB) 

image = np.stack([image_0], axis=0).astype(np.float32)/255.0

image=image.transpose(0,3,1,2)
image=torch.tensor(image)

x_processed, _, _ = SAC_processor(image, bpda=True, shape_completion=True)

image_remove = np.asarray(x_processed[0])
image_remove=image_remove.transpose(1,2,0)

plt.axis("off")
plt.figure(figsize = (20,30))
plt.imshow(image_remove.astype(np.uint8), interpolation="nearest")
plt.show()
```

![test_bike](https://user-images.githubusercontent.com/35388623/211926641-bea61367-d784-4078-900f-8986f99a8376.png)

![test_bike_output](https://user-images.githubusercontent.com/35388623/211926852-649c0b0d-432d-488a-a473-9aad4ade708e.png)

**Comment by Reza99RD** [comment]
Hi,
Thanks a lot for your answer.
I ran the code again with the image you've sent and normalized it to [0,1],
but unfortunately it still doesn't work.
Maybe I'm doing something wrong with the normalization process or with
hyper parameters of "PatchDetector".

My code is in below :

SAC_processor = PatchDetector(3, 1, base_filter=16, square_sizes=[200,150,
100, 75, 50, 25], n_patch=1)
SAC_processor.unet.load_state_dict(torch.load("ckpts/coco_at.pth"
, map_location='cpu'))

image_0 = cv2.imread(
"/content/211926641-bea61367-d784-4078-900f-8986f99a8376.png")
image_0 = cv2.cvtColor(image_0, cv2.COLOR_BGR2RGB)

image_0 = cv2.normalize(image_0, None, alpha=0, beta=1
, norm_type=cv2.NORM_MINMAX, dtype=cv2.CV_32F)

image = np.stack([image_0], axis=0).astype(np.float32)

image=image.transpose(0,3,1,2)
image=torch.tensor(image)

x_processed, _, _ = SAC_processor(image, bpda=True, shape_completion=True)

image_remove = np.asarray(x_processed[0])
image_remove = image_remove.transpose(1,2,0)
image_remove = cv2.normalize(image_remove, None, alpha = 0, beta = 255
, norm_type = cv2.NORM_MINMAX, dtype = cv2.CV_32F)

plt.axis("off")
#plt.figure(figsize = (20,30))
plt.imshow(image_remove.astype(np.uint8), interpolation="nearest")
plt.show()

Thanks in advance

**Comment by joellliu** [comment]
Hi, 

Can you run the code that I used and see if it works for you?

https://github.com/joellliu/SegmentAndComplete/issues/1#issuecomment-1379539736

**Comment by Reza99RD** [comment]
Hi again,
I ran your code just now, and I just add a line to it to convert "image_remove" back to [0,255] for the sake of plot figuration, but again it did not work for me.
I ran this code with google colab on cpu runtime.

my whole code is in below :


```
!pip install kornia
!git clone https://github.com/joellliu/SegmentAndComplete.git

import os
os.chdir('/content/SegmentAndComplete')

```

```
import cv2
import numpy as np
import matplotlib.pyplot as plt
import torch
from patch_detector import PatchDetector

SAC_processor = PatchDetector(3, 1, base_filter=16, square_sizes=[150, 100, 75, 50, 25], n_patch=1)
SAC_processor.unet.load_state_dict(torch.load("ckpts/coco_at.pth", map_location='cpu'))

image_0 = cv2.imread("/content/bike.png")
image_0 = cv2.cvtColor(image_0, cv2.COLOR_BGR2RGB) 

image = np.stack([image_0], axis=0).astype(np.float32)/255.0

image=image.transpose(0,3,1,2)
image=torch.tensor(image)

x_processed, _, _ = SAC_processor(image, bpda=True, shape_completion=True)

image_remove = np.asarray(x_processed[0])
image_remove = image_remove.transpose(1,2,0)

image_remove = image_remove.astype(np.float32)*255.0

plt.axis("off")
plt.figure(figsize = (20,30))
plt.imshow(image_remove.astype(np.uint8), interpolation="nearest")
plt.show()
```

![image_remove](https://user-images.githubusercontent.com/117666885/212165009-b1ae7353-ab23-40ed-94fe-d6ad7a604430.png)

**Comment by joellliu** [comment]
Interesting, not sure what's going on. Can you try this colab notebook that I created?

https://colab.research.google.com/drive/1TdNxkbUUa4ldjJTgPn6LPcmRgF8bWoBo?usp=sharing

**Comment by Reza99RD** [comment]
Now that I tried your colab notebook, it worked.
It also worked with the adversarial patches made by ART Robust Dpatch method (in png format).
Thank you very much, really appreciate it.

![BikeAdv](https://user-images.githubusercontent.com/117666885/212173754-771b1558-b88f-481c-b782-3a292f4344f9.png)

**Comment by joellliu** [comment]
Great! Glad to be of help!
